### PR TITLE
Improved and fixed notifications

### PR DESCRIPTION
**Pull request by PiaFraus**
Improved and fixed notifications

## Description
* Notifications has nothing to do with creating markers. So I moved it out of there to a special processor.
* Notifications don't spam anymore - if more then 3 notification will be spawned in the same update - one general notification will popup.
* Added wildcard(*=all) pokemon to notification selector.
* If something is selected along with the wildcard - it will be excluded from the wildcard. So developers can think of wildcards as inverters of what is selected.
* Pokemon bouncing will happen on click and (by request in comments) when page is just in the background.
* Pokemon bouncing will stop after 2 seconds of notification click.

## Motivation and Context
Notifications is one of the most important feature for me and my friends/colleagues.
Previously we had to select hundreds of them one by one. Now there is a wildcard.
Previously if you refresh a page or just too many pokemons spawned - you will have way too many notifications. Now they are joined together.

## How Has This Been Tested?
Tested locally.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by Toeler** [comment]
> Pokemon bouncing will happen on click and not when page is just in the background.

I rely on this when the map is sitting on my other monitor. Chrome doesn't pop up the toast when you are running a full screen application, so the bounce + ping is what I use.

**Comment by PiaFraus** [comment]
@Toeler I can added a default bounce as well. But for how long?

**Comment by Toeler** [comment]
For the duration that the marker is alive, preferably.

A feature like this is always going to be tough to please everyone.

File: static/map.js [hunk 1]
```diff
@@ -520,20 +530,9 @@ function setupPokemonMarker(item, skipNotification) {
...
 
-    if (notifiedPokemon.indexOf(item.pokemon_id) > -1) {
-        if (!skipNotification) {
-            if (Store.get('playSound')) {
-              audio.play();
-            }
-            sendNotification('A wild ' + item.pokemon_name + ' appeared!', 'Click to load map', 'static/icons/' + item.pokemon_id + '.png', item.latitude, item.longitude);
-        }
-        // Icons still get a bounce, even on redraw
-        marker.setAnimation(google.maps.Animation.BOUNCE);
```
Comment (AHAAAAAAA) [inline static/map.js]: Please include the Bounce animation in your fix.

File: static/map.js [hunk 2]
```diff
@@ -1049,7 +1100,7 @@ $(function () {
 
     // Load pokemon names and populate lists
     $.getJSON("static/locales/pokemon." + language + ".json").done(function(data) {
-        var pokeList = []
+        var pokeList = [ { id: -1, text: '*'}]
```
Comment (AHAAAAAAA) [inline static/map.js]: Can you explain why this is necessary?

Comment (PiaFraus) [inline static/map.js]: * When pokemons are excluded from showing on the map - they already excluded from loading. Most of users I know excluded the "I don't care about pokemons" and had to manually select all other pokemons in the notification selector. This allowed them to just select "all but excluded".
* It also allows to have "Include only selected pokemons", since everything that is selected along with the wildcard will be excluded.

File: static/map.js (hunk 1 above)
Comment (PiaFraus) [inline static/map.js]: @AHAAAAAAA Included. It will still stop bouncing after notification click, so I assume it will have best from both worlds, since people need a way to stop them bouncing.

**Comment by PiaFraus** [comment]
@Toeler I've returned the previous behaviour. But bouncing will still be stopped if user clicks notfication. So I hope it will please the most.

**Comment by hhu94** [comment]
It seems this fix doesn't contain #2017, every time I zoom in/out I get notifications.

**Comment by PiaFraus** [comment]
@hhu94 I agree. I misunderstood with another issue.

**Comment by hhu94** [comment]
A problem I've found is that disabling the Wild Pokemon switch on the sidebar doesn't hide lured pokemon, only normal pokemon.

**Comment by PiaFraus** [comment]
@hhu94 Well I haven't done anything about this switch. But I can actually add the switch check to "isPokemonExcluded" so it would also work as a wildcard.

**Comment by andrewda** [comment]
> Notifications don't spam anymore - if more then 2 notification will be spawned in the same update - one general notification will popup.

Please change this to more than 3.

**Comment by PiaFraus** [comment]
@andrewda Updated to 3.
 That's quite a number of commits. Should I rebase?

File: static/map.js [hunk 3]
```diff
@@ -235,7 +245,7 @@ function initMap() {
               'style_pgo_nl']
         },
     });
-
```
Comment (invisiblek) [inline static/map.js]: while i agree whitespace should be removed, it should be done so in a separate commit

File: static/map.js [hunk 4]
```diff
@@ -531,21 +541,9 @@ function setupPokemonMarker(item, skipNotification, isBounceDisabled) {
...
-            sendNotification('A wild ' + item.pokemon_name + ' appeared!', 'Click to load map', 'static/icons/' + item.pokemon_id + '.png', item.latitude, item.longitude);
-        }
-		if (marker.animationDisabled != true){
-			marker.setAnimation(google.maps.Animation.BOUNCE);	
-		}
-    }
-
     addListeners(marker);
     return marker;
-};
```
Comment (invisiblek) [inline static/map.js]: unrelated to this commit (along with all the other semicolons you removed below)

File: static/map.js (hunk 3 above)
Comment (PiaFraus) [inline static/map.js]: Ok, I'll remove it.
Used to the rules at my work, where if we see style errors, which don't change the behaviour - we just add them with next change instead of creating separate change for them.

File: static/map.js [hunk 5]
```diff
@@ -852,21 +850,84 @@ function processScanned(i, item) {
...
         $.each(result.pokestops, processLuredPokemon);
         $.each(result.gyms, processGyms);
         $.each(result.scanned, processScanned);
-        clearOutOfBoundsMarkers(map_data.pokemons);
-        clearOutOfBoundsMarkers(map_data.lure_pokemons);
-        clearOutOfBoundsMarkers(map_data.gyms);
-        clearOutOfBoundsMarkers(map_data.pokestops);
-        clearOutOfBoundsMarkers(map_data.scanned);
+        // If someone needs this feature - please implement a version that doesn't break notifications
+        // clearOutOfBoundsMarkers(map_data.pokemons);
```
Comment (invisiblek) [inline static/map.js]: i'd rather we leave this in personally, it helps quite a bit with larger maps

Comment (PiaFraus) [inline static/map.js]: It forces markers to be deleted and this introduced a bug, that when you move a map - all notifications are spawned again. What problem this supposed to fix? Memory issues?
Personally if I am on map and move left-right - I would prefer to immidately see existed markers, since they already loaded instead of spawning a new one and notifying me about it. I need notification for something I couldn't see. Something that is currently out of the bounds.

Comment (invisiblek) [inline static/map.js]: well, when we fetch new data, we're already leaving out anything not on our map, so when we pan, we see what was there before, sure, but  we wont see anything new until it refreshes again

i guess what i'm saying is if we're going this direction (not pruning out of bounds markers) we should just pull all the data too (if the reason for doing the former is for quicker markers when panning)

Comment (invisiblek) [inline static/map.js]: regardless, it feels to me this wildcard thing is a new feature that i don't quite understand, and i'm not sure we want to introduce it for 2.1.0 (which https://github.com/AHAAAAAAA/PokemonGo-Map/issues/1951 needs to be fixed for)

**Comment by hhu94** [comment]
I really like the wildcard feature, it allows me to exclude the pokemon I'm not interested and receive notifications for all the other pokemon that aren't in the ignore list.

Also, this PR is precisely meant to fix #1951. As far as I have tested it, it does indeed fix it.

File: static/map.js (hunk 5 above)
Comment (BillyWM) [inline static/map.js]: Let's not break one feature to fix another

> It forces markers to be deleted and this introduced a bug, that when you move a map - all notifications are spawned again. 

That's a more systemic problem with the code. Markers and notifications shouldn't be tied together to begin with. setupPokemonMarker shouldn't have this responsibility. It isn't setupPokemonMarkerAndSendNotifications. It should do one thing....set up markers.

There should probably be one function that just decides whether to send a notification. processPokemons is calling setupPokemonMarker. It should also call this new function e.g. sendNotifications, and that function can run logic to decide whether a notification is warranted (is this pokemon in the viewable area? Has a notification already been sent for this spawn? Is it part of the exclusion list. Etc)

>What problem this supposed to fix? Memory issues?

Memory and performance. If you think a function like that is useless, you shouldn't just yank it out of the code. Removing something unrelated that has been in there a while should be an isolated PR of its own, open for discussion.

Comment (hhu94) [inline static/map.js]: I really like the wildcard feature, it allows me to exclude the pokemon I'm not interested and receive notifications for all the other pokemon that aren't in the ignore list.

Also, this PR is precisely meant to fix #1951. As far as I have tested it, it does indeed fix it.

Comment (PiaFraus) [inline static/map.js]: I agree. Personally I'd just get all the pokemon data for my needs.
But I am fine with this middle-ground solution: 
Good sides:
-  it doesn't load everything for those who have multiple scanners on big areas
-  it also doesn't delete them on pan for people like me and doesn't force second notification.

Bad sides:
- people who constantly pan a lot for areas with new pokemons might have slow expirience.
- people who forgot to reset their view will not be notified about out-of-bound pokemons

Comment (PiaFraus) [inline static/map.js]: @BillyWM, 

> Let's not break one feature to fix another

Agree. This lines were added and broke the notification feature, so I am removing the code, that broke a feature.

Comment (PiaFraus) [inline static/map.js]: @invisiblek 
Actually wildcard is extremely simple. It just adds * to the list of pokemons. I think it's common notation, that star means "include everything".

Comment (BillyWM) [inline static/map.js]: >Memory and performance.
>Only for those who constantly pans. If it's out of bounds - it's not loaded in the first place.

Panning is kind of an important feature of the map! If the new notification code has to hamper panning, it's broken, and it should not be merged until it's fixed.

Please either restore the clearOutOfBoundsMarkers calls or rework the function so that it doesn't interfere with notifications.

What is it doing that interferes anyhow?

>And clearOutOfBoundsMarkers is not called clearOutOfBoundsMarkersAndAlsoAllThePokemons,

Pokemon, stops, and gyms on the map are markers. I don't get what your point is.

This is called the Single Responsibility Principle. I didn't make this idea up myself.

Comment (PiaFraus) [inline static/map.js]: @BillyWM

> Panning is kind of an important feature of the map! If the new notification code has to hamper panning, it's broken, and it should not be merged until it's fixed.

Notifications of pokemons is an important feature of an app which is about telling people about pokemons.
If this panning code broke the already implemented notification system, which I just fix - it was broken and should be removed as a fix which my change is about.

> Please either restore the clearOutOfBoundsMarkers calls or rework the function so that it doesn't interfere with notifications.

@invisiblek is fixing this method I think. I might wait for him to finish and see later. 
As right now with this change - everything works perfectly fine. All my friends and colleagues are much happier with my fix. I am not in rush.

> Pokemon, stops, and gyms on the map are markers. 

No. No they are not. They are map_data. - Data from the server.
For simplicity instead of creating markers in separated container, code ammends this data and ADDS to data also a marker property. Addition to a data. But map_data.pokemons - is not a list of markers. So
clearOutOfBoundsMarkers actually violates the separation of concerns principle and deletes not just markers, but all the received data.

Comment (BillyWM) [inline static/map.js]: >clearOutOfBoundsMarkers actually violates the separation of concerns principle and deletes not just markers, but all the received data.

Yes, that's wrong, but it doesn't *need* to be doing that. And nor do notifications need to be ripping out performance features.

Comment (invisiblek) [inline static/map.js]: https://github.com/AHAAAAAAA/PokemonGo-Map/pull/2214

The best I could come up with (working on the current constraints) was that.

Comment (BillyWM) [inline static/map.js]: > It might a little bit slow down for people who constantly look at new pokemons.

So, with this notification code, adding markers to the map triggers notifications.

So....no markers can be removed, because they would re-trigger notifications.

So if someone scans a wide or dense area (San Francisco, NYC, etc) they just have to live with tons of markers degrading performance, even if they zoom in only to their neighborhood or college campus?

Comment (BillyWM) [inline static/map.js]: I see this as analogous to game engine rendering. If there's a ton of entities, you don't blindly run all their draw logic all the time. You only have a draw list of those on screen.

If markers are on the map, Google Maps spends time managing them. Tons of markers = bad performance, regardless of whether they're currently rendered.

Managing the viewable area is a necessary step for performance.

Comment (PiaFraus) [inline static/map.js]: @BillyWM,
> So....no markers can be removed until they're expired, because they would re-trigger notifications
No. No data should be removed if it's used by other features. It was broken by this function. I removed the broken code. If someone wants to properly implement the perfomance boost by not destroying important features - good! But this is about fixing previously broken code.

> I  see this as analogous to game engine rendering. If there's a ton of entities, you don't blindly run all their draw logic all the time. You only have a draw list of those on screen.
Brilliant example. If you don't see where enemy is - your code shouldn't delete all info about this enemy. It shouldn't draw it - yes. But enemy can interact with you. Shoot at you even when it's not at your sight. If you submit a code, that deletes all enemies as soon as you don't see them - you remove their change and ask them to properly implement what they wanted.

Comment (PiaFraus) [inline static/map.js]: I might spend this night trying to properly implemented what they wanted to do. I'll see if I have time.

Comment (jjbott) [inline static/map.js]: At least for the clearOutOfBoundsMarkers fix, I think we're looking for something simple like : https://github.com/jjbott/PokemonGo-Map/commit/f841bc613109323200b68f7371d4b4253d23aa8a . Don't delete everything, just show/hide based on the visible map

**Comment by xzyfer** [comment]
>For the duration that the marker is alive, preferably.

Please no. The bouncing is wildly distracting. At the very least we should be able to disable the bounce. The desktop notification is plenty for a lot of people.

**Comment by Slivo-fr** [comment]
+1

Maybe give the user the choice of bounce or not, and duration.
I like to have the marker bounce it's whole lifetime on my side.

Agree with OP on clearOutOfBoundsMarkers side. It sounds like a broken feature.

**Comment by PiaFraus** [comment]
@AHAAAAAAA 
I've just noticed your comments in #1322 and #1415 :
Wildcard('*') is essentially "All". I can rename it if you think it will be more clear.
And I've added it to both selectors and to just Notify as I think it's good if similar UI elements are as similar to each other as it could be.
Plus there are a couple of request to invert this fields and additional feature of wildcard allows that automatically.

**Comment by PiaFraus** [comment]
@xzyfer @Slivo-fr Well currently bouncing is stopped if:
a) You click on any notification
b) You click on bouncing marker
c) You change zoom to remove bouncing for all of them
I don't really want to introduce new UI elements so may be we can think of something? But if you think it will really help -I'll do it.

**Comment by Slivo-fr** [comment]
Current behavior is fine for me, was just throwing some ideas :)

**Comment by invisiblek** [comment]
moving this to 2.2.0 as its now become a feature and we're trying to get 2.1.0 finished

**Comment by jnovack** [comment]
> That's quite a number of commits. Should I rebase?

@PiaFraus always rebase and squash.

**Comment by hhu94** [comment]
Can't wait for this to be merged. It's essential to the notification system.

**Comment by PiaFraus** [comment]
@hhu94 Feel free to take ownership of it. I am not dealing with grunt.

**Comment by invisiblek** [comment]
Rebased here: https://github.com/invisiblek/PokemonGo-Map/commit/4acc1bd47d89bf4c748609113731970ef4dd4c83

But I must admit, I still don't get the point of the wildcards. What possible reason could you have to hide ALL the Pokemon at once or display all notifications at once?

**Comment by hhu94** [comment]
@invisiblek How I use it is I hide all the pokemon I don't want to see and just enter the star in the notifications list. That way I get notified of any pokemon that appears as long as it's not in my ignore list.

**Comment by invisiblek** [comment]
Wouldn't it make more sense to have it be a toggle instead of adding it to the list? My thought would be toggling it on would grey out the list box (like it does when you turn the "Follow location" toggle on) and then act like the behavior you're describing. I still don't see the point of having it in the "Hide Pokemon" box though...

**Comment by PiaFraus** [comment]
@invisiblek Main reason for me - simpler UI. 
At home I want to see everything around and be notified about everything but a few common ones.
Exclude pokemons=empty; Notify = * + Zubat/Goblat/...
At work I want to see only a bunch of rare, if I look at the map when I have time. I want to be notified only for super rare, since it can interrupt my workflow
Exclude pokemons = * + Eevee/...; Notify = Dratini/...

All of this options and more with just two UI elements.
Without wildcard there would probably be twice as many ui elements.

**Comment by Slivo-fr** [comment]
The idea of toggles for wildcard make more sense imo. 
If this particular feature is to be discuted maybe you should split it in a second pull request ?

**Comment by Slivo-fr** [comment]
Why the hell closing it ? it's still bug atm and I'm waiting for those improvements for a week now !

**Comment by invisiblek** [comment]
PiaFraus has no intention of rebasing the patch. If you'd like to do so, feel free to submit a new pull request.
